doc/developer/design/20220314_linear_or_eventual.md

```diff
@@ -0,0 +1,115 @@
+
+We want to fix the lack of any kind of consistency for "unmaterialized sources" AND offer _linearizable_ reads for all types of sources. In short, we want the following
+table to represent all possible interactions with sources in `materialized`:
+
+
+|              | unmaterialized | materialized   |
+| ------------ | -------------- | -------------- |
+| Enventual    | Not available  | Default        |
+| Linearizable | Default        | `LINEARIZABLE` |
+
+Not offering non-_linearizable_ reads for reads interacting with unmaterialized sources simplifies the implementation, and clarifies the user experience of
+working with unmaterialized sources
+
+## Goals
+
+1. Offer _linearizable_ reads in as many cases as possible. `materialized` is ultimately built to support this!
+2. Offer eventual consistency by default (keeping the same behavior we currently have in `materialized`) for materialized sources
+
+## Non-Goals
+
+Alter the default experience of using `materialized` (yet).
+
+## Description
+
+### Background
+
+Sources currently produce messages that contain offsets. Those offsets are written down, associated with a timestamp, as a _timestamp binding_.
+When you do a peek (read) against some source. A recent timestamp is chosen, the index is consulted for data that is valid in respect to that timestamp
+and the _timestamp bindings_.
+
+### Linearizable reads
+In the case of `LINEARIZABLE` read (see below for how to express this in sql), peeks will be augemented to work as follows
+
+1. Before issuing the peek, we will first ask each source involved in the peek: "what is your max offset"
+2. We will wait until a max offset is available at some timestamp `t` (i.e., written down as a timestamp binding)
+3. And then issue the peek for timestamp `t`.
+
+In the case of "unmaterialized" sources, the temporary index will be created at the beginning of this process and not removed until the end.
+
+### Eventually consistent reads
+This is the default behavior of peeks, and will be augmented as follows:
+
+1. Before issuing the peek, we will look for the latest available timestamp for all involved sources
+2. And then issue the peek at that timestamp
+
+This is the standard `since` and `uppser` logic that already exists in `materialized`. Additionally, "unmaterialized"
+sources won't have access to do this. A query involving this
+
+In the case of "unmaterialized" sources, there are 2 important notes:
+- the temporary index will be created at the beginning of this process and not removed until the end.
+- When an unmaterialized source is created, a background special _index_ will be created, which will progress
+the source and produce only offsets that are bound to timestamps. This allows EC reads against unmaterialized sources
+to consistently obtain new data.
+
+### Platform considerations
+
+In the world of Platform, when performing a _linearizable_ read, we need to be specific about what we mean by
+steps 1&2. The way these will operate is:
```

Comment:
Ideally, we could define things in terms of what happens in platform, rather than "what happens in pre-platform code" and then diff that with platform changes.